src/components/tl/cuda/tl_cuda_team.c

```diff
@@ -0,0 +1,303 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_cuda.h"
+#include "tl_cuda_coll.h"
+#include "tl_cuda_topo.h"
+#include "core/ucc_team.h"
+#include "coll_score/ucc_coll_score.h"
+#include "utils/arch/cpu.h"
+#include <sys/shm.h>
+
+UCC_CLASS_INIT_FUNC(ucc_tl_cuda_team_t, ucc_base_context_t *tl_context,
+                    const ucc_base_team_params_t *params)
+{
+    ucc_tl_cuda_context_t *ctx =
+        ucc_derived_of(tl_context, ucc_tl_cuda_context_t);
+    ucc_tl_cuda_lib_t     *lib =
+        ucc_derived_of(tl_context->lib, ucc_tl_cuda_lib_t);
+    ucc_tl_cuda_shm_barrier_t *bar;
+    ucc_status_t status;
+    int shm_id, i, j;
+    size_t ctrl_size;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_team_t, &ctx->super, params);
+
+    self->oob    = params->params.oob;
+    self->stream = NULL;
+    self->topo   = NULL;
+    if (UCC_TL_TEAM_SIZE(self) > UCC_TL_CUDA_MAX_PEERS) {
+        tl_info(tl_context->lib, "team size is too large, max supported %d",
+                UCC_TL_CUDA_MAX_PEERS);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    if (!ucc_topo_is_single_node(params->team->topo)) {
+        tl_info(tl_context->lib, "multinode team is not supported");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    ctrl_size = (sizeof(ucc_tl_cuda_sync_t) + sizeof(ucc_tl_cuda_sync_data_t) *
+                (UCC_TL_TEAM_SIZE(self) - 1)) * UCC_TL_TEAM_SIZE(self) *
+                lib->cfg.max_concurrent +
+                sizeof(ucc_tl_cuda_shm_barrier_t) * lib->cfg.max_concurrent +
+                sizeof(ucc_tl_cuda_sync_state_t) * lib->cfg.max_concurrent;
+
+    shm_id = -1;
+    self->sync = (void*)-1;
+    if (UCC_TL_TEAM_RANK(self) == 0) {
+        shm_id = shmget(IPC_PRIVATE, ctrl_size, IPC_CREAT | 0666);
+        if (shm_id < 0) {
+            tl_error(tl_context->lib, "failed to shmget with IPC_PRIVATE, "
+                     "size %zd, IPC_CREAT errno: %d(%s)", ctrl_size,
+                     errno, strerror(errno));
+            status = UCC_ERR_NO_MEMORY;
+            goto exit_err;
+        }
+        self->sync = shmat(shm_id, NULL, 0);
```

Comment:
if shmget /shmat fails then rank 0 will return with error and never go to oob -> dead lock on other ranks? lets use negative shm_id and make rank 0 always go to allgather to process global error state.

Reply:
done